### ISSUE TITLE
debug: debugging session terminates before the debugee finishes its exit job

### ISSUE THREAD

**Issue by ztibeike** [golang/vscode-go#2137]
I want to capture the OS signal and do some exit jobs when I terminate the vscode golang debugger.

I have code as below:
```go
sigalChan := make(chan os.Signal, 1)
signal.Notify(sigalChan, syscall.SIGINT, syscall.SIGTERM)
<-sigalChan
doSomeJobs()
```
but it doesn't work. Anyone can tell me how to figure it out? Maybe the signal type is not SIGINT or SIGTERM?

**Comment by findleyr** [comment]
I see you also asked at https://stackoverflow.com/questions/71602708/what-os-signal-does-the-vscode-go-debugger-send-to-delve-dap-when-i-terminate-de, and I suspect that the advice there is accurate.

@polinasok or @suzmue might know if it is possible to catch the debugger exit.

**Comment by ztibeike** [comment]
> I see you also asked at https://stackoverflow.com/questions/71602708/what-os-signal-does-the-vscode-go-debugger-send-to-delve-dap-when-i-terminate-de, and I suspect that the advice there is accurate.
> 
> @polinasok or @suzmue might know if it is possible to catch the debugger exit.

Yeah I also asked at stackoverflow and got some advice which however do not solve my problem.  I still want to know if there's any way to figure it out.

**Comment by polinasok** [comment]
@ztibeike In theory you need to start a headless dlv server in a terminal and connect to it with a client. In that terminal, the debuggee will be foregrounded, so it can receive the signals. It looks like when that happens vscode-go both with legacy and with dlv-dap adapter does the wrong thing. Execution is paused on an error and the rest of the code doesn't get executed.  I need to debug a bit more to understand why this happens. In the meantime, you can use dlv terminal client instead:

```
$ dlv debug hello.go --headless --listen=:54321
API server listening at: [::]:54321
2022-03-25T09:35:09-07:00 warning layer=rpc Listening for remote connections (connections are not authenticated nor encrypted)
debugserver-@(#)PROGRAM:LLDB  PROJECT:lldb-1205.0.27
 for x86_64.
Got a connection, launched process /Users/polina/go/src/hello/__debug_bin (pid = 57085).

...waiting for signal
^C
!!! got signal !!!
```
```
$ dlv connect :54321
Type 'help' for list of commands.
(dlv) c
Stopped at: 0x30b3ad0
=>   1:	no source available
Command failed: EOF
(dlv) exit
```

**Comment by ztibeike** [comment]
> @ztibeike In theory you need to start a headless dlv server in a terminal and connect to it with a client. In that terminal, the debuggee will be foregrounded, so it can receive the signals. It looks like when that happens vscode-go both with legacy and with dlv-dap adapter does the wrong thing. Execution is paused on an error and the rest of the code doesn't get executed. I need to debug a bit more to understand why this happens. In the meantime, you can use dlv terminal client instead:
> 
> ```
> $ dlv debug hello.go --headless --listen=:54321
> API server listening at: [::]:54321
> 2022-03-25T09:35:09-07:00 warning layer=rpc Listening for remote connections (connections are not authenticated nor encrypted)
> debugserver-@(#)PROGRAM:LLDB  PROJECT:lldb-1205.0.27
>  for x86_64.
> Got a connection, launched process /Users/polina/go/src/hello/__debug_bin (pid = 57085).
> 
> ...waiting for signal
> ^C
> !!! got signal !!!
> ```
> 
> ```
> $ dlv connect :54321
> Type 'help' for list of commands.
> (dlv) c
> Stopped at: 0x30b3ad0
> =>   1:	no source available
> Command failed: EOF
> (dlv) exit
> ```

Thanks a lot! It works! Instead of manually running "dlv debug ...", I configure a pre-launch-task in tasks.json to start the debug mode of delve server and then attach to it, I think it's more efficient. But there's still a little problem as you got, I can catch the exit and run the rest of code by pressing "ctrl+c", but can't  to do so by clicking the UI button.

**Comment by ztibeike** [comment]
> @polinasok Thanks a lot! It works! Instead of manually running "dlv debug ...", I configure a pre-launch-task in tasks.json to start the debug mode of delve server as below.
> 
> ```json
> // tasks.json
> {
>   "version": "2.0.0",
>   "tasks": [
>     {
>       "label": "start dlv",
>       "type": "shell",
>       "command": "/root/go/bin/dlv",
>       "isBackground": true,
>       "args": [
>         "debug",
>         "${workspaceFolder}/main.go",
>         "--headless",
>         "--listen=:54321",
>         "--check-go-version=false",
>         "--api-version=2"
>       ]
>     }
>   ]
> }
> ```
> 
> Then attach to it . I think it's more efficient. But there's still a little problem as you got, I can catch the exit and run the rest of code by pressing "ctrl+c", but can't to do so by clicking the UI button. The error log is:
> 
> ```
> 2022-03-26T14:24:43+08:00 error layer=rpc writing response:write tcp 127.0.0.1:54321->127.0.0.1:40676: use of closed network connection
> ```
> 
> I think the debugger just shut down it's listening port to close the connection, then delve server crashes, the golang program exit along with the delve.

I've found that such pre-launch-task is unnecessary. I can get the same situation when I set "console" to "integratedTerminal" in launch.json .

**Comment by hyangah** [comment]
Good to hear that `console` attribute with "integratedTerminal" or "externalTerminal" helps this. 
Closing.

**Comment by hyangah** [comment]
@polinasok pointed out the issue is not resolved. We need to investigate what causes`dlv dap` to terminate before the debugee to complete the exit job.

**Comment by davidxiao** [comment]
<img width="1302" alt="Screen Shot 2022-09-09 at 5 10 56 pm" src="https://user-images.githubusercontent.com/4176280/189292406-0ae5019e-9845-4ec2-9fc7-afdb955980cf.png">

it doesn't work for me with the setting in screenshot, 

any advice?

Thanks

**Comment by ztibeike** [comment]
> 
My configuration is totally the same as yours and it works. Can you or how do you reproduce the issue?

**Comment by ztibeike** [comment]
> <img alt="Screen Shot 2022-09-09 at 5 10 56 pm" width="1302" src="https://user-images.githubusercontent.com/4176280/189292406-0ae5019e-9845-4ec2-9fc7-afdb955980cf.png">
> 
> it doesn't work for me with the setting in screenshot, getting `error layer=dap runtime error: EOF`
> 
> any advice?
> 
> Thanks

My configuration is totally the same as yours and it works. Can you or how do you reproduce the issue?

**Comment by davidxiao** [comment]
yeah, i always got the error

**Comment by firelizzard18** [comment]
I think ultimately the issue is that delve terminates (SIGKILL) the process instead of asking it to stop (SIGINT, SIGTERM). I think "allow a debug session to be gracefully terminated" fully covers this issue.
